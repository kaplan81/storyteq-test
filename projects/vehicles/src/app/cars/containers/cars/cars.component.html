<div class="cars">
  <h2 class="cars__heading">Buy your car with us :)</h2>
  @if (vehicleState().loaded) {
    <ul class="cars__list">
      @for (car of cars(); track $index) {
        <li class="cars__list__item">
          <img class="cars__list__photo" [src]="car.media[mediaApectRatio()].url" alt="" />
          <div class="cars__list__photo__info">
            <div class="cars__list__photo__info__name">
              {{ car.name.toUpperCase() }}
            </div>
            <div class="cars__list__photo__info__price">
              {{ car.detail?.price }}
            </div>
            <div class="cars__list__photo__info__description">
              {{ car.detail?.description }}
            </div>
          </div>
        </li>
      }
    </ul>
  } @else {
    <teq-spinner></teq-spinner>
  }
</div>
