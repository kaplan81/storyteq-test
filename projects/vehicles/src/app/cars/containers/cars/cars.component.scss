@use 'mixins';

.cars {
  &__list {
    display: flex;
    flex-direction: column;
    @include mixins.bigger-than(md) {
      flex-direction: row;
      flex-wrap: wrap;
    }
    &__item {
      display: flex;
      @include mixins.bigger-than(md) {
        flex-basis: 50%;
        flex-direction: column;
      }
      @include mixins.bigger-than(lg) {
        flex-basis: 25%;
      }
      &__photo {
        width: 150px;
        @include mixins.bigger-than(md) {
          width: 100%;
        }
      }
      &__info {
        flex-grow: 1;
        border: 1px solid #eee;
        padding: 20px;
        &__name {
          font-weight: bold;
          margin-bottom: 7px;
        }
        &__price {
          margin-bottom: 12px;
          color: #666;
        }
        &__description {
          color: #666;
        }
      }
    }
  }
}
