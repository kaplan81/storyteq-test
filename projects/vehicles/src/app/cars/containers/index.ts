export * from '@vehicles/cars/containers/cars/cars.component';
