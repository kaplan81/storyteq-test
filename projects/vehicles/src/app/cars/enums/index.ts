export * from '@vehicles/cars/enums/media.enum';
