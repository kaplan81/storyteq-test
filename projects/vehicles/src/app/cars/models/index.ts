export * from '@vehicles/cars/models/vehicle-state.model';
export * from '@vehicles/cars/models/vehicle.model';
