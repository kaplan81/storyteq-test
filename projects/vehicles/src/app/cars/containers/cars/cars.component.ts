import { BreakpointObserver, BreakpointState } from '@angular/cdk/layout';
import {
  ChangeDetectionStrategy,
  Component,
  Signal,
  WritableSignal,
  computed,
  inject,
  signal,
} from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { SpinnerComponent } from '@vehicles/app/components';
import { Breakpoint, RoutedComponents } from '@vehicles/app/enums';
import { TitleMixin, emptyBase } from '@vehicles/app/mixins';
import { MediaAspectRatio } from '@vehicles/cars/enums';
import { VechicleState, VehicleEntity } from '@vehicles/cars/models';
import { VehicleStateService } from '@vehicles/cars/services/vehicle-state/vehicle-state.service';
import { VehicleService } from '@vehicles/cars/services/vehicle/vehicle.service';

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  imports: [SpinnerComponent],
  standalone: true,
  styleUrl: './cars.component.scss',
  templateUrl: './cars.component.html',
})
export class CarsComponent extends TitleMixin(
  emptyBase,
  RoutedComponents[RoutedComponents.cars],
  true,
) {
  #breakpointObserver = inject(BreakpointObserver);
  #vehicleService = inject(VehicleService);
  #vehicleStateService = inject(VehicleStateService);
  cars: Signal<VehicleEntity[]> = this.#vehicleStateService.getSortedVehicleEntities();
  vehicleState: Signal<VechicleState> = computed<VechicleState>(() =>
    this.#vehicleStateService.state(),
  );
  mediaApectRatio: WritableSignal<MediaAspectRatio> = signal<MediaAspectRatio>(
    MediaAspectRatio.mobile,
  );

  constructor() {
    super();
    this.#breakpointObserver
      .observe([`(min-width: ${Breakpoint.md})`])
      .pipe(takeUntilDestroyed())
      .subscribe((breakpointState: BreakpointState) => {
        if (breakpointState.matches) {
          console.log(`Viewport width is ${Breakpoint.md} or greater!`);
          this.mediaApectRatio.set(MediaAspectRatio.nonMobile);
        } else {
          console.log(`Viewport width is less than ${Breakpoint.md}!`);
          this.mediaApectRatio.set(MediaAspectRatio.mobile);
        }
      });
    this.#vehicleService.getCars().pipe(takeUntilDestroyed()).subscribe();
  }

  getPrice(car: VehicleEntity): string {
    let price = '';
    if (car.detail !== null) {
      if (car.detail.price !== undefined) {
        price = car.detail.price;
      }
    }
    return price.length > 0 ? `From ${price}` : 'No price available';
  }
}
