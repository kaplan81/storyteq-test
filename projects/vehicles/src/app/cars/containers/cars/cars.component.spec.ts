import { provideHttpClient } from '@angular/common/http';
import { provideHttpClientTesting } from '@angular/common/http/testing';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { CarsComponent } from './cars.component';

describe('CarsComponent', () => {
  let fixture: ComponentFixture<CarsComponent>;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [CarsComponent],
      providers: [provideHttpClient(), provideHttpClientTesting()],
    }).compileComponents();
    fixture = TestBed.createComponent(CarsComponent);
  });

  it('should match snapshot', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
