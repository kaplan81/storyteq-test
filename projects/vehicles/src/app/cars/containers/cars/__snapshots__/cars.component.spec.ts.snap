// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarsComponent should match snapshot 1`] = `
<ng-component
  cars={[Function Function]}
  mediaApectRatio={[Function Function]}
  title={[Function String]}
  vehicleState={[Function Function]}
>
  <div
    class="cars"
  >
    <teq-spinner>
      <span
        class="spinner"
      />
    </teq-spinner>
  </div>
</ng-component>
`;
